File: leetcode/tree/2519_Count_the_Number_of_K_Big_Indices/codes/fentree.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class FenwickTree {
public:
    vector<int> tree;
    FenwickTree(int size) {
        tree.resize(size + 2);
    }

    void update(int i, int delta) {
        while (i < tree.size()) {
            tree[i] += delta;
            i += (i & -i);
        }
    }

    int query(int i) {
        int res = 0;
        while (i > 0) {
            res += tree[i];
            i -= (i & -i);
        }
        return res;
    }
};

class Solution {
public:
    int kBigIndices(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> sorted = nums;
        sort(sorted.begin(), sorted.end());
        unordered_map<int, int> compress;
        int id = 1;
        for (int num : sorted)
            if (compress.find(num) == compress.end())
                compress[num] = id++;

        vector<int> comp(n);
        for (int i = 0; i < n; ++i)
            comp[i] = compress[nums[i]];

        vector<int> left(n), right(n);
        FenwickTree bit(id);

        for (int i = 0; i < n; ++i) {
            left[i] = bit.query(comp[i] - 1);
            bit.update(comp[i], 1);
        }

        bit = FenwickTree(id);
        for (int i = n - 1; i >= 0; --i) {
            right[i] = bit.query(comp[i] - 1);
            bit.update(comp[i], 1);
        }

        int count = 0;
        for (int i = 0; i < n; ++i) {
            if (left[i] >= k && right[i] >= k)
                ++count;
        }
        return count;
    }
};

File: leetcode/tree/2519_Count_the_Number_of_K_Big_Indices/codes/fentree.py
from bisect import bisect_left
from typing import List

class FenwickTree:
    def __init__(self, size):
        self.tree = [0] * (size + 2)

    def update(self, i, delta):
        while i < len(self.tree):
            self.tree[i] += delta
            i += (i & -i)

    def query(self, i):
        result = 0
        while i > 0:
            result += self.tree[i]
            i -= (i & -i)
        return result

class Solution:
    def kBigIndices(self, nums: List[int], k: int) -> int:
        sorted_unique = sorted(set(nums))
        compressed = [bisect_left(sorted_unique, x) + 1 for x in nums] 
        n = len(nums)
        left_smaller = [0 for _ in range(n)]
        right_smaller = [0 for _ in range(n)]

        bit = FenwickTree(len(sorted_unique))
        for i in range(n):
            left_smaller[i] = bit.query(compressed[i] - 1)
            bit.update(compressed[i], 1)

        bit = FenwickTree(len(sorted_unique))
        for i in reversed(range(n)):
            right_smaller[i] = bit.query(compressed[i] - 1)
            bit.update(compressed[i], 1)

        return sum(1 for i in range(n) if left_smaller[i] >= k and right_smaller[i] >= k)
